py3status/modules/yubikey.py

```diff
@@ -0,0 +1,167 @@
+                    self.py3.update()
+
+            def run(this):
+                inotify = inotify_simple.INotify()
+                watches = []
+                for path in self.gpg_check_watch_paths:
+                    if os.path.isfile(path):
+                        watch = inotify.add_watch(path,
+                                                  inotify_simple.flags.OPEN)
+                        watches.append(watch)
+
+                while not self.killed.is_set():
+                    for event in inotify.read():
+                        # e.g. ssh doesn't start immediately, try several seconds to be sure
+                        for i in range(20):
+                            with this._check_card_status() as check:
+                                try:
+                                    # if this doesn't return very quickly, touch is pending
+                                    check.communicate(
+                                        timeout=self.gpg_check_timeout)
+                                    time.sleep(0.25)
+                                except subprocess.TimeoutExpired:
+                                    this._set_pending(True)
+                                    os.killpg(check.pid, signal.SIGINT)
+                                    check.communicate()
+                                    # wait until device is touched
+                                    with this._check_card_status() as wait:
+                                        wait.communicate()
+                                        this._set_pending(False)
+                                        break
+
+                        # ignore all events caused by operations above
+                        for _ in inotify.read():
+                            pass
+
+                for watch in watches:
+                    inotify.rm_watch(watch)
+
+        class SudoThread(threading.Thread):
+            def _set_pending(this, new_value):
+                self.pending_sudo = new_value
+                if not self.pending_gpg:
+                    self.py3.update()
+
+            def run(this):
+                inotify = inotify_simple.INotify()
+                watches = []
+                for path in self.sudo_check_watch_paths:
+                    if os.path.isfile(path):
+                        watch = inotify.add_watch(path,
+                                                  inotify_simple.flags.OPEN)
+                        watches.append(watch)
+
+                while not self.killed.is_set():
+                    for event in inotify.read():
+                        this._set_pending(True)
+                        for event in inotify.read():
+                            this._set_pending(False)
+
+                for watch in watches:
+                    inotify.rm_watch(watch)
+
+        if len(self.gpg_check_watch_paths) > 0:
+            GpgThread().start()
+
+        if len(self.sudo_check_watch_paths) > 0:
```

Comment:
(Nit). `if len(self.sudo_check_watch_paths):` would work okay here too.

Reply:
yeah, but I like it explicit